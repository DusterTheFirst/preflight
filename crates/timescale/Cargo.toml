[package]
name = "timescale"
version = "0.1.0"
authors = ["Zachary Kohnen <14093962+DusterTheFirst@users.noreply.github.com>"]
edition = "2018"

[dependencies]
timescale_derive = { path = "./derive" }
serde = "1.0"
lerp = { branch = "derive_macro", git = "https://github.com/DusterTheFirst/lerp-rs/", features = ["derive"] }
