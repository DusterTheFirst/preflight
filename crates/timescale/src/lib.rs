pub use lerp::Lerp;
use serde::Serialize;

/// Trait to allow data points to be expanded into their timescaled/serializable counterpart
/// by appending the time
pub trait ToTimescale {
    type Timescale: Serialize;

    /// Create the equivalent timescaled data point by appending the time
    /// to this data point
    fn with_time(self, time: f64) -> Self::Timescale;
}

pub enum TimescaleData<D> {
    Interpolation { low: D, high: D, percent: f64 },
    HighSaturation(D),
    LowSaturation(D),
}

/// Trait to allow for linear interpolation through a static timescale lookup table for smoother lookups
pub trait TimescaleDataTable {
    type Datapoint: Lerp<f64>;

    /// Get a data from the lookup table with all metadata attached. Prefer to use the `get_lerp` methods.
    fn get(time: f64) -> TimescaleData<Self::Datapoint>;

    /// Get data from the lookup table, linear interpolating between points
    /// if the time given is between 2 points on the table, fully saturating
    /// if the time is outside of the table's range
    fn get_lerp(time: f64) -> Self::Datapoint {
        match Self::get(time) {
            TimescaleData::HighSaturation(d) | TimescaleData::LowSaturation(d) => d,
            TimescaleData::Interpolation { low, high, percent } => low.lerp(high, percent),
        }
    }
}

#[allow(unused_imports)]
#[macro_use]
extern crate timescale_derive;

#[doc(hidden)]
pub use timescale_derive::*;

// #[cfg(test)]
mod test {
    use lerp::Lerp;

    use super::{TimescaleData, TimescaleDataTable};

    struct DataTable;
    impl TimescaleDataTable for DataTable {
        type Datapoint = f64;

        fn get(time: f64) -> TimescaleData<Self::Datapoint> {
            match time {
                _ if time <= 0.0 => TimescaleData::LowSaturation(0.0),
                // 0.0 => TimescaleData::Literal(0.0),
                _ if time >= 0.0 && time < 0.5 => {
                    let time_high = 0.5;
                    let time_low = 0.0;

                    TimescaleData::Interpolation {
                        high: 100.0,
                        low: 0.0,
                        percent: (time - time_low) / (time_high - time_low),
                    }
                }
                // 0.5 => TimescaleData::Literal(100.0),
                _ if time >= 0.5 && time < 1.0 => {
                    let time_high = 1.0;
                    let time_low = 0.5;

                    TimescaleData::Interpolation {
                        high: 300.0,
                        low: 100.0,
                        percent: (time - time_low) / (time_high - time_low),
                    }
                }
                // 1.0 => TimescaleData::Literal(300.0),
                _ if time >= 1.0 => TimescaleData::HighSaturation(300.0),
                _ => unreachable!(),
            }
        }
    }

    #[test]
    fn datatable_1_unit() {
        assert_eq!(DataTable::get_lerp(-0.1), 0.0);
        // assert_eq!(DataTable::get_quantized(-0.1), None);

        assert_eq!(DataTable::get_lerp(0.0), 0.0);
        // assert_eq!(DataTable::get_quantized(0.0), Some(0.0));

        assert_eq!(DataTable::get_lerp(0.1), 20.0);
        // assert_eq!(DataTable::get_quantized(0.1), None);

        assert_eq!(DataTable::get_lerp(0.5), 100.0);
        // assert_eq!(DataTable::get_quantized(0.5), Some(100.0));

        assert_eq!(DataTable::get_lerp(0.6), 140.0);
        // assert_eq!(DataTable::get_quantized(0.6), None);

        assert_eq!(DataTable::get_lerp(1.0), 300.0);
        // assert_eq!(DataTable::get_quantized(1.0), Some(300.0));

        assert_eq!(DataTable::get_lerp(1.1), 300.0);
        // assert_eq!(DataTable::get_quantized(1.1), None);
    }

    struct DataTable2;
    struct Data2(f64, f64);

    impl TimescaleDataTable for DataTable2 {
        type Datapoint = Data2;

        fn get(time: f64) -> TimescaleData<Self::Datapoint> {
            match time {
                _ if time <= 0.0 => TimescaleData::LowSaturation(0.0),
                // 0.0 => TimescaleData::Literal(0.0),
                _ if time >= 0.0 && time < 0.5 => {
                    let time_high = 0.5;
                    let time_low = 0.0;

                    TimescaleData::Interpolation {
                        high: 100.0,
                        low: 0.0,
                        percent: (time - time_low) / (time_high - time_low),
                    }
                }
                // 0.5 => TimescaleData::Literal(100.0),
                _ if time >= 0.5 && time < 1.0 => {
                    let time_high = 1.0;
                    let time_low = 0.5;

                    TimescaleData::Interpolation {
                        high: 300.0,
                        low: 100.0,
                        percent: (time - time_low) / (time_high - time_low),
                    }
                }
                // 1.0 => TimescaleData::Literal(300.0),
                _ if time >= 1.0 => TimescaleData::HighSaturation(300.0),
                _ => unreachable!(),
            }
        }
    }


    #[test]
    fn datatable_2_units() {
        assert_eq!(DataTable::get_lerp(-0.1), 0.0);
        // assert_eq!(DataTable::get_quantized(-0.1), None);

        assert_eq!(DataTable::get_lerp(0.0), 0.0);
        // assert_eq!(DataTable::get_quantized(0.0), Some(0.0));

        assert_eq!(DataTable::get_lerp(0.1), 20.0);
        // assert_eq!(DataTable::get_quantized(0.1), None);

        assert_eq!(DataTable::get_lerp(0.5), 100.0);
        // assert_eq!(DataTable::get_quantized(0.5), Some(100.0));

        assert_eq!(DataTable::get_lerp(0.6), 140.0);
        // assert_eq!(DataTable::get_quantized(0.6), None);

        assert_eq!(DataTable::get_lerp(1.0), 300.0);
        // assert_eq!(DataTable::get_quantized(1.0), Some(300.0));

        assert_eq!(DataTable::get_lerp(1.1), 300.0);
        // assert_eq!(DataTable::get_quantized(1.1), None);
    }
}
