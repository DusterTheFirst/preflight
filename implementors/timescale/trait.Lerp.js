(function() {var implementors = {};
implementors["timescale"] = [];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()