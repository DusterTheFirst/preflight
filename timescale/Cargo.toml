[package]
name = "timescale"
version = "0.1.0"
authors = ["Zachary Kohnen <14093962+DusterTheFirst@users.noreply.github.com>"]
edition = "2018"
description = "A crate to work with timescale data and csv files"

[dependencies]
timescale_macros = { path = "./macros" }
serde = "1.0"
lerp = { branch = "derive_macro", git = "https://github.com/DusterTheFirst/lerp-rs/", features = ["derive"] }

[dev-dependencies]
trybuild = "1.0"
