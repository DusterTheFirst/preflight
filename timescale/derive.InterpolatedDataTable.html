<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `InterpolatedDataTable` derive in crate `timescale`."><meta name="keywords" content="rust, rustlang, rust-lang, InterpolatedDataTable"><title>timescale::InterpolatedDataTable - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc derive"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../timescale/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><div class="sidebar-elems"><p class="location"><a href="index.html">timescale</a></p><script>window.sidebarCurrent = {name: "InterpolatedDataTable", ty: "derive", relpath: ""};</script><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../settings.html"><img src="../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/timescale_macros/lib.rs.html#57" title="goto source code">[src]</a></span><span class="in-band">Derive Macro <a href="index.html">timescale</a>::<wbr><a class="derive" href="">InterpolatedDataTable</a></span></h1><pre class="rust derive">#[derive(InterpolatedDataTable)]
{
    // Attributes available to this derive:
    #[table]
}</pre><div class="docblock"><p>Load a csv file into a static data table with support for linearly interpolating
data points between the discreet time intervals</p>
<p>This derive macro goes hand and hand with the <a href="../timescale/derive.InterpolatedData.html" title="InterpolatedData"><code>InterpolatedData</code></a> derive
macro. Data loaded into this table needs to be decorated with the aforementioned
<a href="../timescale/derive.InterpolatedData.html" title="InterpolatedData"><code>InterpolatedData</code></a> derive macro so that their structure can be available to
this macro at build time.</p>
<h1 id="arguments" class="section-header"><a href="#arguments">Arguments</a></h1>
<p>This macro requires 2 arguments.</p>
<ul>
<li><code>st</code> describes the structure (that derives <a href="../timescale/derive.InterpolatedData.html" title="InterpolatedData"><code>InterpolatedData</code></a>) to represent
the timescale data.</li>
<li><code>file</code> describes the csv file to read in the timescale data from</li>
</ul>
<h1 id="example" class="section-header"><a href="#example">Example</a></h1>
<p>See the data in this <a href="https://github.com/DusterTheFirst/preflight/blob/master/assets/motors/Estes_A8.csv">csv file</a> for context for the following example</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">timescale</span>::{<span class="ident">InterpolatedData</span>, <span class="ident">InterpolatedDataTable</span>};

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Debug</span>, <span class="ident">Lerp</span>, <span class="ident">InterpolatedData</span>)]</span>
<span class="kw">pub</span> <span class="kw">struct</span> <span class="ident">RocketEngine</span> {
    <span class="attribute">#[<span class="ident">data</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;Thrust (N)&quot;</span>)]</span>
    <span class="kw">pub</span> <span class="ident">thrust</span>: <span class="ident">f64</span>,
}

<span class="doccomment">/// The thrust curve of an Estes A8 rocket motor</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">InterpolatedDataTable</span>)]</span>
<span class="attribute">#[<span class="ident">table</span>(<span class="ident">file</span> <span class="op">=</span> <span class="string">&quot;../assets/motors/Estes_A8.csv&quot;</span>, <span class="ident">st</span> <span class="op">=</span> <span class="string">&quot;RocketEngine&quot;</span>)]</span>
<span class="kw">pub</span> <span class="kw">struct</span> <span class="ident">EstesA8</span>;

<span class="kw">fn</span> <span class="ident">main</span>() {
    <span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">EstesA8</span>::<span class="ident">get</span>(<span class="number">0.35</span>).<span class="ident">thrust</span>, <span class="number">3.813</span>); <span class="comment">// Exact value from data</span>
    <span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">EstesA8</span>::<span class="ident">get</span>(<span class="number">0.4</span>).<span class="ident">thrust</span>, <span class="number">3.9468823529411763</span>); <span class="comment">// Linear interpolated estimate</span>
    <span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">EstesA8</span>::<span class="ident">get</span>(<span class="number">1.0</span>).<span class="ident">thrust</span>, <span class="number">0.0</span>); <span class="comment">// Saturated at 0</span>
}</pre></div>
</div></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../";window.currentCrate = "timescale";</script><script src="../main.js"></script><script defer src="../search-index.js"></script></body></html>