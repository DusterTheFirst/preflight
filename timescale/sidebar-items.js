initSidebarItems({"derive":[["InterpolatedData","Load the layout of a structure to be consumed by [`InterpolatedDataTable`]"],["InterpolatedDataTable","Load a csv file into a static data table with support for linearly interpolating data points between the discreet time intervals"],["Lerp",""],["ToTimescale","Automatically derive the `ToTimescale` trait and generate the expanded timescale structure."]],"enum":[["InterpolatedDataPoint","A constant datapoint that could be interpolated or saturated"]],"trait":[["InterpolatedDataTable","Trait to allow for linear interpolation through a static timescale table"],["Lerp","Types which are amenable to linear interpolation and extrapolation."],["ToTimescale","Trait to allow data points to be expanded into their timescaled and serializable counterpart by appending the time"]]});