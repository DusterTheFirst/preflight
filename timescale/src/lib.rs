use serde::Serialize;

pub trait Timescale {
    type Timescaled: Serialize;

    fn with_time(self, time: f64) -> Self::Timescaled;
}

#[allow(unused_imports)]
#[macro_use]
extern crate timescale_derive;

#[doc(hidden)]
pub use timescale_derive::*;
