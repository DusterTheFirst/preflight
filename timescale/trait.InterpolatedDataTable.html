<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `InterpolatedDataTable` trait in crate `timescale`."><meta name="keywords" content="rust, rustlang, rust-lang, InterpolatedDataTable"><title>timescale::InterpolatedDataTable - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc trait"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../timescale/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><p class="location">Trait InterpolatedDataTable</p><div class="sidebar-elems"><div class="block items"><a class="sidebar-title" href="#associated-types">Associated Types</a><div class="sidebar-links"><a href="#associatedtype.Datapoint">Datapoint</a><a href="#associatedtype.Time">Time</a></div><a class="sidebar-title" href="#associated-const">Associated Constants</a><div class="sidebar-links"><a href="#associatedconstant.MAX">MAX</a><a href="#associatedconstant.MIN">MIN</a></div><a class="sidebar-title" href="#required-methods">Required Methods</a><div class="sidebar-links"><a href="#tymethod.get_raw">get_raw</a></div><a class="sidebar-title" href="#provided-methods">Provided Methods</a><div class="sidebar-links"><a href="#method.get">get</a></div><a class="sidebar-title" href="#implementors">Implementors</a></div><p class="location"><a href="index.html">timescale</a></p><script>window.sidebarCurrent = {name: "InterpolatedDataTable", ty: "trait", relpath: ""};</script><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../settings.html"><img src="../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/timescale/lib.rs.html#42-70" title="goto source code">[src]</a></span><span class="in-band">Trait <a href="index.html">timescale</a>::<wbr><a class="trait" href="">InterpolatedDataTable</a></span></h1><div class="docblock type-decl hidden-by-usual-hider"><pre class="rust trait">pub trait InterpolatedDataTable: <a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Send.html" title="trait core::marker::Send">Send</a> + <a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Sync.html" title="trait core::marker::Sync">Sync</a> + 'static {
    type <a href="#associatedtype.Datapoint" class="type">Datapoint</a>: <a class="trait" href="../timescale/trait.Lerp.html" title="trait timescale::Lerp">Lerp</a>&lt;Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a>&gt;;
    type <a href="#associatedtype.Time" class="type">Time</a>: <a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Copy.html" title="trait core::marker::Copy">Copy</a>;

    const <a href="#associatedconstant.MIN" class="constant"><b>MIN</b></a>: Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a>;
    const <a href="#associatedconstant.MAX" class="constant"><b>MAX</b></a>: Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a>;

    fn <a href="#tymethod.get_raw" class="fnname">get_raw</a>(<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;time: Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a><br>&nbsp;&nbsp;&nbsp;&nbsp;) -&gt; <a class="enum" href="../timescale/enum.InterpolatedDataPoint.html" title="enum timescale::InterpolatedDataPoint">InterpolatedDataPoint</a>&lt;Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Datapoint" title="type timescale::InterpolatedDataTable::Datapoint">Datapoint</a>, Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a>&gt;;

    fn <a href="#method.get" class="fnname">get</a>(time: Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a>) -&gt; Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Datapoint" title="type timescale::InterpolatedDataTable::Datapoint">Datapoint</a> { ... }
}</pre></div><div class="docblock"><p>Trait to allow for linear interpolation through a static timescale table</p>
</div><h2 id="associated-types" class="small-section-header">Associated Types<a href="#associated-types" class="anchor"></a></h2><div class="methods"><h3 id="associatedtype.Datapoint" class="method"><code>type <a href="#associatedtype.Datapoint" class="type">Datapoint</a>: <a class="trait" href="../timescale/trait.Lerp.html" title="trait timescale::Lerp">Lerp</a>&lt;Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a>&gt;</code></h3><div class="docblock"><p>The datapoint type</p>
</div><h3 id="associatedtype.Time" class="method"><code>type <a href="#associatedtype.Time" class="type">Time</a>: <a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Copy.html" title="trait core::marker::Copy">Copy</a></code></h3><div class="docblock"><p>The scalar used for timing, Only f64 or f32 are supported</p>
</div></div><span class="loading-content">Loading content...</span><h2 id="associated-const" class="small-section-header">Associated Constants<a href="#associated-const" class="anchor"></a></h2><div class="methods"><h3 id="associatedconstant.MIN" class="method"><code>const <a href="#associatedconstant.MIN" class="constant"><b>MIN</b></a>: Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a></code></h3><div class="docblock"><p>The minimum time value that is in the dataset</p>
</div><h3 id="associatedconstant.MAX" class="method"><code>const <a href="#associatedconstant.MAX" class="constant"><b>MAX</b></a>: Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a></code></h3><div class="docblock"><p>The maximum time value that is in the dataset</p>
</div></div><span class="loading-content">Loading content...</span><h2 id="required-methods" class="small-section-header">Required methods<a href="#required-methods" class="anchor"></a></h2><div class="methods"><h3 id="tymethod.get_raw" class="method"><code>fn <a href="#tymethod.get_raw" class="fnname">get_raw</a>(<br>&nbsp;&nbsp;&nbsp;&nbsp;time: Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a><br>) -&gt; <a class="enum" href="../timescale/enum.InterpolatedDataPoint.html" title="enum timescale::InterpolatedDataPoint">InterpolatedDataPoint</a>&lt;Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Datapoint" title="type timescale::InterpolatedDataTable::Datapoint">Datapoint</a>, Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a>&gt;</code></h3><div class="docblock"><p>Get a data from the lookup table with all metadata attached. Prefer to use the <code>get</code> method, for
this method is an implementation detail</p>
</div></div><span class="loading-content">Loading content...</span><h2 id="provided-methods" class="small-section-header">Provided methods<a href="#provided-methods" class="anchor"></a></h2><div class="methods"><h3 id="method.get" class="method"><code>fn <a href="#method.get" class="fnname">get</a>(time: Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Time" title="type timescale::InterpolatedDataTable::Time">Time</a>) -&gt; Self::<a class="type" href="../timescale/trait.InterpolatedDataTable.html#associatedtype.Datapoint" title="type timescale::InterpolatedDataTable::Datapoint">Datapoint</a></code></h3><div class="docblock"><p>Get data from the lookup table, linear interpolating between points
if the time given is between 2 points on the table, fully saturating
if the time is outside of the table's range</p>
</div></div><span class="loading-content">Loading content...</span><h2 id="implementors" class="small-section-header">Implementors<a href="#implementors" class="anchor"></a></h2><div class="item-list" id="implementors-list"></div><span class="loading-content">Loading content...</span><script type="text/javascript" src="../implementors/timescale/trait.InterpolatedDataTable.js" async></script></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../";window.currentCrate = "timescale";</script><script src="../main.js"></script><script defer src="../search-index.js"></script></body></html>