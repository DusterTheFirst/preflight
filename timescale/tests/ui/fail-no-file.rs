use lerp::Lerp;
use timescale::{InterpolatedData, InterpolatedDataTable};

#[derive(Lerp, InterpolatedData)]
pub struct RocketEngine {
    #[data(rename = "Thrust (N)")]
    pub thrust: f64,
}

#[derive(InterpolatedDataTable)]
#[table(file = "../../../assets/motors/ghajkhgdaskjdgh.csv", st = "RocketEngine")]
pub struct EstesC6;

fn main() {}