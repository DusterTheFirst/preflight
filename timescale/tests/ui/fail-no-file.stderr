error: `$WORKSPACE/target/tests/timescale/../../../assets/motors/ghajkhgdaskjdgh.csv` does not exist
  --> $DIR/fail-no-file.rs:11:16
   |
11 | #[table(file = "../../../assets/motors/ghajkhgdaskjdgh.csv", st = "RocketEngine")]
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
